### PR TITLE
feat: add data shuffle and random seed option

### PR DESCRIPTION
**Pull request by ZhiyuLi-Nvidia**
# What does this PR do ?

As titled.

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA/NeMo-RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA/NeMo-RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA/NeMo-RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_rl/algorithms/dpo.py [hunk 1]
```diff
@@ -153,7 +153,7 @@ def setup(
     train_dataloader = StatefulDataLoader(
         train_dataset,
         batch_size=policy_config["train_global_batch_size"],
-        shuffle=True,
+        shuffle=data_config["shuffle"],
```
Comment (SahilJain314) [inline nemo_rl/algorithms/dpo.py:155]: Can we add seeding if we're shuffling? I don't want subsequent runs to randomly repeat data.

Comment (ZhiyuLi-Nvidia) [reply nemo_rl/algorithms/dpo.py:155]: Good point. Is that a good idea to set global random seed like 

https://github.com/huggingface/transformers/blob/v4.51.3/src/transformers/trainer_utils.py#L92-L125

**Comment by SahilJain314** [comment]
@ZhiyuLi-Nvidia I realized we never actually merged this. Can you take a look at seeding whenever you can?

File: nemo_rl/algorithms/dpo.py (hunk 1 above)
Comment (ZhiyuLi-Nvidia) [reply nemo_rl/algorithms/dpo.py:155]: Yeap. 
* Added seed in dataset setup 
* randomness in dataloader is controlled by global seed state from `set_seed`.

Verified the data is reproducible.

Comment (SahilJain314) [reply nemo_rl/algorithms/dpo.py:155]: Can you verify that data is _not_ reproduced across restarts?

Comment (SahilJain314) [reply nemo_rl/algorithms/dpo.py:155]: (maybe add a unit test for this)

Comment (ZhiyuLi-Nvidia) [reply nemo_rl/algorithms/dpo.py:155]: Added a unit test for data shuffling reproducibility. 
We should expect the same dataloader in ordering with the same seed.

File: tests/unit/data/test_data_shuffle_reproducity.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    "chat_template": "default",
+}
+
+SEED = 42
+BATCH_SIZE = 4
+MAX_SEQ_LENGTH = 128
+MAX_BATCHES_TO_TEST = 10
+
+
+def create_dataloader(seed: int = SEED) -> StatefulDataLoader:
```
Comment (SahilJain314) [inline tests/unit/data/test_data_shuffle_reproducity.py]: Thanks for this test. I'd also really like to see a test that upon restart, the same data isn't repeated (even with seeding). The checkpointed state should take precedence.

Even better would be to see that the data loaded across restarts is the same as without restarting, which should all be controllable by seed.

Comment (ZhiyuLi-Nvidia) [reply tests/unit/data/test_data_shuffle_reproducity.py]: This test is to mimic restarting the run, say creating a dataset and dataloader from a seed and if we can have the same batch outputs accordingly.

I have updated the tests for better readability.

Comment (ZhiyuLi-Nvidia) [reply tests/unit/data/test_data_shuffle_reproducity.py]: >  I want to see the following:
dataloader with seed
run for 10 batches
new dataloader with seed
run for 5 batches
checkpoint
restore with seed
run for 5 batches
the second dataloader's batches even with restore should equal the first one

Just added `test_data_shuffle_reproducity_from_continue` for test reproducity from a continued experiment.

**Review by SahilJain314 — APPROVED** [review]

**Comment by ashors1** [comment]
LGTM from the SFT and DPO side, but just wondering -- does this change impact convergence of our recipes (esp. grpo ones)? Wonderinfg whether we need to adjust our test suite at all

**Comment by ZhiyuLi-Nvidia** [comment]
@terrykong @parthchadha 
Feel free to let me know if you have any comments

**Review by terrykong — DISMISSED** [review]

**Review by terrykong — APPROVED** [review]
